### PR TITLE
add `create_column` to Registry API

### PR DESCRIPTION
**Pull request by ayushkamat**

### CODE REVIEW

File: latch/registry/table.py [hunk 1]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
 
         mutations.append(res)
 
+    # upsert column
+
+    def create_column(
+        self,
+        key: str,
+        type: Union[Type[RegistryPrimitivePythonValue], str, List[str]],
```
Comment (AidanAbd) [inline latch/registry/table.py:568]: This is bad -- there should be explicit classes which indicate the types

Union[Type[RegistryPrimitivePythonValue], RegistryPrimitiveLinkType, RegistryPrimitiveEnumType]

File: latch/registry/table.py [hunk 2]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
 
         mutations.append(res)
 
+    # upsert column
+
+    def create_column(
```
Comment (AidanAbd) [inline latch/registry/table.py:565]: Confusing if this can be used to update enums as well.

File: latch/registry/table.py [hunk 3]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+        if type is LatchFile:
+            registry_type = {"primitive": "blob"}
+        if type is LatchDir:
+            registry_type = {"primitive": "blob", "metadata": {"nodeType": "dir"}}
+
+        if isinstance(type, str):
+            registry_type = {"primitive": "link", "experimentId": type}
+        if isinstance(type, List):
+            if all(isinstance(t, str) for t in type):
+                registry_type = {"primitive": "enum", "members": type}
```
Comment (AidanAbd) [inline latch/registry/table.py:621]: Check against a class which can hold the reference informations
ex
```
RegistryPrimitiveLinkType(reference="...")
RegistryPrimitiveEnumType(members=[...])
```

File: latch/registry/table.py [hunk 4]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+
+        if registry_type is None:
+            raise InvalidColumnError(
+                key,
+                str(type),
+                f"python type {type} could not be converted to registry type",
+            )
+
+        if allow_multiple_entries:
+            registry_type = {"array": registry_type}
```
Comment (AidanAbd) [inline latch/registry/table.py:631]: I would move the allow_multiple_entries check here and check the constructed type.

**Review by AidanAbd — CHANGES_REQUESTED** [review]

File: latch/registry/table.py (hunk 2 above)
Comment (maximsmol) [inline latch/registry/table.py:565]:
```suggestion
    def upsert_column(
```

File: latch/registry/table.py [hunk 5]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+                as the id of Table and will produce a link column. If a list of
+                strings is provided, they are treated as members of an Enum and
+                will produce a select column.
+            allow_multiple: Whether or not to allow multiple entries in the
+                new column. Only valid if `type` is a file, directory, or link.
+            required: Whether or not a value is required in the new column.
+
+        """
+
+        if allow_multiple_entries:
```
Comment (maximsmol) [inline latch/registry/table.py:589]: rework as discussed

File: latch/registry/table.py [hunk 6]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+
+    def create_column(
+        self,
+        key: str,
+        type: Union[Type[RegistryPrimitivePythonValue], str, List[str]],
+        *,
+        allow_multiple_entries: bool = False,
+        required: bool = False,
+    ):
+        """Create a column.
```
Comment (maximsmol) [inline latch/registry/table.py:573]:
```suggestion
        """Update or create a column.
```

File: latch/registry/table.py [hunk 7]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+
+        A transport error is thrown if a column of the same name already exists.
+
+        Args:
+            name: New column name.
```
Comment (maximsmol) [inline latch/registry/table.py:578]:
```suggestion
            key: Identifier of the new column.
```

File: latch/registry/table.py [hunk 8]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+            type: Type of new column. If a string is provided, it is treated
```
Comment (maximsmol) [inline latch/registry/table.py:579]:
```suggestion
            type: Type of the new column. If a string is provided, it is treated
```
+ update the rest (tho it should be self-describing based on the function type

File: latch/registry/table.py [hunk 9]
```diff
@@ -541,6 +560,119 @@ def _add_record_deletes_selection(
...
+                as the id of Table and will produce a link column. If a list of
+                strings is provided, they are treated as members of an Enum and
+                will produce a select column.
+            allow_multiple: Whether or not to allow multiple entries in the
+                new column. Only valid if `type` is a file, directory, or link.
+            required: Whether or not a value is required in the new column.
```
Comment (maximsmol) [inline latch/registry/table.py:585]:
```suggestion
            required: If true, records without a value for this column are considered invalid. Note that an explicit `None` value is different from a missing/empty value.
```

**Review by maximsmol — CHANGES_REQUESTED** [review]

**Review by AidanAbd — APPROVED** [review]

**Pull request merged**